#                              Exploring Hackers News Posts
![Image](https://s3.amazonaws.com/dq-content/354/hacker_news.jpg)


In this project, we'll compare two different types of posts from [Hacker News][1], a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either `Ask HN` or `Show HN`.

Users submit `Ask HN` posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit `Show HN` posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

- Do `Ask H` or `Show HN` receive more comments on average?
- Do posts created at a certain time receive more comments on average?

It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.You can find the original data set [here](https://www.kaggle.com/hacker-news/hacker-news-posts).

The columns are as folowing:

- `id:` the unique identifier from Hacker News for the post
- `title:` the title of the post
- `url:` the URL that the posts links to, if the post has a URL
- `num_points:` the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
- `num_comments:` the number of comments on the post
- `author:` the username of the person who submitted the post
- `created_at:` the date and time of the post's submission


Let's start by importing the libraries we need and reading the dataset into a list of lists.

[1]:https://news.ycombinator.com/


In [1]:
from csv import reader

def open_dataset(file_name):
    '''Opens a file and returns the dataset

    Args:
        path (str): The location of a CSV file.

    Returns:
        list of lists: each row is list inside a list 
    '''  
    opened_file = open(file_name,  encoding='utf8')
    read_file = reader(opened_file)
    data = list(read_file)
    return data


def get_header(data):
    '''Get the header from a dataset
    
    Args:
        data(list): list of lists to get header
    
    Returns:
        list: first list in lists or the header
    '''
    return data[0]


def get_data_no_header(data):
    '''Get the data without the header row
    Args:
        data(list): list of lists to drop header
    
    Returns:
        list of lists: the rows without the header row
    '''
    return data[1:]


hn_data=open_dataset("data/HN_posts_year_to_Sep_26_2016.csv")

def explore_data(dataset_has_no_header, start, end, rows_and_columns=False):
    '''Prints rows of data from strat to end index and the number of rows and columns
    if the rows_and_columns flag is set to True
    '''
    dataset_slice=dataset_has_no_header[start:end]
    for row in dataset_slice:
        print(f'\n{row}\n') #prints row followed by a new line
        
    if rows_and_columns: 
        #checks the flag and prints # of rows and columns
        print(f'Number of rows:{len(dataset_has_no_header)}')
        print(f'Number of columns:{len(dataset_has_no_header[0])}\n')


In [2]:
hn=open_dataset('data\hacker_news.csv')
headers=get_header(hn_data)
print(f'Header:\n{headers}\n')
hn=get_data_no_header(hn)
explore_data(hn,0,5,True)

Header:
['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']

Num

# Extracting Ask HN and Show HN Posts
First, we'll identify posts that begin with either `Ask HN` or `Show HN` and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.


In [3]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts=[]
show_posts=[]
other_posts=[]

for row in  hn:
    title = row[1] #column 2 is the title
    title=title.lower()
    
    if title.lower().startswith("show hn"):
        show_posts.append(row)
    elif title.lower().startswith("ask hn"):
        ask_posts.append(row)
    else:
        other_posts.append(row)
        
ask_post_len=len(ask_posts)
show_posts_len=len(show_posts)
other_posts_len=len(other_posts)
print (f'Number of different kind of posts:\n  {ask_post_len = }\n  {show_posts_len = }\n  {other_posts_len =}')
                                      

Number of different kind of posts:
  ask_post_len = 1744
  show_posts_len = 1162
  other_posts_len =17194


# Calculating the Average Number of Comments for Ask HN and Show HN Posts
Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [4]:
total_ask_comments=0
for row in ask_posts:
    num_comments=int(row[4]) # column 5 is the number of posts
    total_ask_comments+=num_comments
avg_ask_comments=total_ask_comments/len(ask_posts)


print(f'{total_ask_comments = }')
print(f'{avg_ask_comments = :.2f}')
print(f'Average Rounded Ask Comments = {round(avg_ask_comments)}\n')

total_show_comments=0
for row in show_posts:
    num_comments=int(row[4]) # column 5 is the number of posts
    total_show_comments+=num_comments
    avg_show_comments=total_show_comments/len(show_posts)
    
    
print(f'{"-"*30}\n')
print(f'{total_show_comments = }')
print(f'{avg_show_comments = :.2f}')
print(f'Average Rounded Show Comments = {round(avg_show_comments)}\n')
print(f'{"-"*30}\n')

# total_other_comments=0
# for row in other_posts:
#     num_comments=int(row[4])
#     total_other_comments+=num_comments
#     avg_other_comments=total_other_comments/len(other_posts)

# print(f'{"-"*30}\n')
# print(f'{total_other_comments = }\n')
# print(f'{avg_other_comments= :2f}\n')
# print(f'Average Rounded other Comments = {round(avg_other_comments)}\n')
 

total_ask_comments = 24483
avg_ask_comments = 14.04
Average Rounded Ask Comments = 14

------------------------------

total_show_comments = 11988
avg_show_comments = 10.32
Average Rounded Show Comments = 10

------------------------------



On average, `ask posts` in our sample receive approximately `14` comments, whereas `show posts` receive approximately `10`. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

We will exract the created_at and number of coments from each row and create a `result_list`

then going through that list we create to dictionaries as folowing :

- `counts_by_hour:` contains the number of ask posts created during each hour of the day.
- `comments_by_hour:` contains the corresponding number of comments ask posts created at each hour received.

In [5]:
import datetime as dt

result_list=[]

for row in ask_posts:
    created_at=row[6] # columns 7 is created_at (%m/%d/%Y %H:%M)
    num_comments=int(row[4])  # column 5 is the number of posts
    result_list.append(
        (created_at,num_comments)
    )

counts_by_hour , comments_by_hour={},{}
date_format='%m/%d/%Y %H:%M'

for row in result_list:
    hour=row[0] # first element of tuple - date/time
    date_object=dt.datetime.strptime(hour, date_format)
    hour=f'{date_object:%H}'                             #date_object.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=row[1] # second element of tuple - number of posts
    else:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]
    
print(f'{counts_by_hour = }')
print(f'{comments_by_hour = }')

counts_by_hour = {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}
comments_by_hour = {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


# Calculating the Average Number of Comments for Ask HN Posts by Hour
Next, we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.
- `avg_by_hour:` is a list of lists containing the hours during which posts were created and the average number of comments those posts received.


In [6]:
avg_by_hour=[]

for hour in counts_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])

print(f'Average number of comments by hour sorted, descending):\n{avg_by_hour}\n')


Average number of comments by hour sorted, descending):
[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]



# Sorting on number of comment and Printing Values from a List of Lists

Creat a list with where order is swapped:
- `swap_avg_by_hour`

Then assign the sorted resuluts from highest average number of comments to lowest by using the sorted method and assign the results to :

- `sorted_swap`

In [7]:
swap_avg_by_hour=[]

for element in avg_by_hour:
    swap_avg_by_hour.append([element[1],element[0]])
    
sorted_swap=sorted(swap_avg_by_hour, reverse=True)
print(f'{sorted_swap = }')   

sorted_swap = [[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


Then we use the f-strings  to print the hour and average in the following format: `15:00: 38.59 average comments per post`.

In [8]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for Ask Posts Comments")

for avg, hr in sorted_swap[:5]:
    print(f'{hr}:00: {avg: .2f} average comments per post')
    

#the old way :
# print("Top 5 Hours for Ask Posts Comments")
# template="{}:00: {:.2f} average comments per post"
# for avg, hr in sorted_swap[:5]:
#     print(template.format(hr,avg))
    

Top 5 Hours for Ask Posts Comments
15:00:  38.59 average comments per post
02:00:  23.81 average comments per post
20:00:  21.52 average comments per post
16:00:  16.80 average comments per post
21:00:  16.01 average comments per post


The hour that receives the most comments per post on average is `15:00`, with an `average of 38.59 comments per post`. There's about a `60% increase` in the number of comments between the hours with the highest and second highest average number of comments.

# Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00.

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 received the most comments on average.